runtime/README.md

```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+| chksum                      | u32      |                                                          |
+| context size                | u32      | size of context                                          |
+| context                     | u8[...]  | This MUST come from the output of the `CM_ECDH_GENERATE` |
+| incoming exchange data size | u32      |                                                          |
+| incoming exchange data      | u8[...]  | the other side's public point                            |
+
+*Table: `CM_ECDH_FINISH` output arguments*
+| **Name**    | **Type** | **Description**                  |
+| ----------- | -------- | -------------------------------- |
+| chksum      | u32      |                                  |
+| fips_status | u32      | FIPS approved or an error        |
+| output CMID | u8[32]   | Output CMID of the shared secret |
+
+
+### CM_RANDOM_STIR
+
+This allows additional entropy to be added to the underlying deterministic random bit generator.
+
+Command Code: `0x434D_5253` ("CMRS")
+
+*Table: `CM_RANDOM_STIR` input arguments*
+
+| **Name**   | **Type** | **Description** |
+| ---------- | -------- | --------------- |
+| chksum     | u32      |                 |
+| input size | u32      | size of input   |
+| input      | u8[...]  |                 |
+
+*Table: `CM_RANDOM_STIR` output arguments*
+| **Name**    | **Type** | **Description**              |
+| ----------- | -------- | ---------------------------- |
+| chksum      | u32      |                              |
+| fips_status | u32      | FIPS approved or an error    |
+
+### CM_RANDOM_GENERATE
+
+This generates random bytes that are returned from the internal RNG.
+
+Command Code: `0x434D_5247` ("CMRG")
+
+*Table: `CM_RANDOM_GENERATE` input arguments*
+
+| **Name**            | **Type** | **Description** |
+| ------------------- | -------- | --------------- |
+| chksum              | u32      |                 |
+| data size to return | u32      |                 |
+
+
+*Table: `CM_RANDOM_GENERATE` output arguments*
+| **Name**    | **Type** | **Description**           |
+| ----------- | -------- | ------------------------- |
+| chksum      | u32      |                           |
+| fips_status | u32      | FIPS approved or an error |
+| output size | u32      | size of output            |
+| output      | u8[...]  |                           |
+
+### CM_IMPORT
+
+Imports the specified data and returns a CMID for it.
+
+Usage information is required so that the data can be verified and used appropriately.
+
+Command Code: `0x434D_494D` ("CMIM")
+
+*Table: `CM_IMPORT` input arguments*
+
+| **Name**   | **Type** | **Description**                            |
+| ---------- | -------- | ------------------------------------------ |
+| chksum     | u32      |                                            |
+| usage      | u32      | Tag to specify how the data can be used    |
+|            |          | 0 - Can be used for HMAC, HKDF, AES        |
```

Comment:
I feel like we could be more prescriptive. Something like,

```
0 - HMAC-SHA2-256
1 - HMAC-SHA2-384
2 - HMAC-SHA2-512
3 - HKDF-SHA2-256
4 - HKDF-SHA2-384
5 - HKDF-SHA2-512
6 - AES-256-GCM
```
Then each key can only be used for a specific purpose, promoting good hygiene. CMIDs returned from `CM_ECDH_FINISH` could be implicitly tagged with type `HKDF-SHA2-384`. Callers could explicitly pass a value from this list as an arg to `CM_HKDF_EXPAND` (which could potentially replace the "Output length" arg).

Reply:
I'm about 50% agreed, except that it is really hard to differentiate between keys for HMAC and HKDF, and TLS isn't required to use HKDF (it's only a `SHOULD`).

Reply:
A TLS implementation would know whether a given key is about to be used for HMAC or HKDF, so the caller should be able to specify at each step. To support either case for a key that comes out of `CM_HKDF_EXPAND`, we could allow the caller to specify the purpose as an input arg to that function.